### DISCUSSION TITLE
Update failed - server data missing files

### DISCUSSION THREAD

**Discussion by Android3000** [mbround18/valheim-docker#216] [Q&A]
Hello,

First of all, thanks for the awesome Docker image. This has helped make things so much easier to manage!

This could totally be my fault, but posting here in case. During the server update process, there was an error in the update job as noted in the logs below. It appears that this may have led to some incomplete server data, as the server was now missing the `globalgamemanagers` file in `valheim_server_Data`. The `globalgamemanagers.assets` and `globalgamemanagers.assets.resS` were present, but not the core file. This was leading to a crash on startup. Deleting the server files and letting it download everything fresh again worked just fine to restore the server.

For transparency, I'm doing some server automation and for the time being was manually invoking auto_update.sh via `docker exec` (versus relying on the cron job) -- however, I had tested this a few times successfully so I don't know that my method of running it was at fault, but I wanted to be clear about that in case what I was doing could have been responsible.

Here are the relevant server logs for reference:

```
Connecting anonymously to Steam Public...Logged in OK
Waiting for user info...OK
Update state (0x3) reconfiguring, progress: 0.00 (0 / 0)
Update state (0x3) reconfiguring, progress: 0.00 (0 / 0)
Update state (0x61) downloading, progress: 0.10 (143636 / 149317150)
Update state (0x61) downloading, progress: 0.10 (143636 / 149317150)
Update state (0x61) downloading, progress: 0.48 (712468 / 149317150)
Update state (0x61) downloading, progress: 0.48 (712468 / 149317150)
Update state (0x61) downloading, progress: 0.48 (712468 / 149317150)
Update state (0x61) downloading, progress: 0.48 (712468 / 149317150)
Update state (0x61) downloading, progress: 0.48 (712468 / 149317150)
Update state (0x61) downloading, progress: 0.48 (712468 / 149317150)
Update state (0x61) downloading, progress: 37.29 (55678198 / 149317150)
Update state (0x61) downloading, progress: 100.00 (149317150 / 149317150)
Update state (0x301) stopping, progress: 0.00 (0 / 149317150)
Error! App '896660' state is 0x626 after update job.
[ODIN][INFO] - Exited with status code: 8

...

[ODIN][INFO] - Executable: /home/steam/valheim/valheim_server.x86_64
[ODIN][INFO] - Launching Command...
[ODIN][INFO] - Checking for BepInEx Environment...
[ODIN][INFO] - Everything looks good! Running normally!
[ODIN][INFO] - Success, daemonized
Found path: /home/steam/valheim/valheim_server.x86_64
Mono path[0] = '/home/steam/valheim/valheim_server_Data/Managed'
Mono config path = '/home/steam/valheim/valheim_server_Data/MonoBleedingEdge/etc'
Preloaded 'libsteam_api.so'
No GlobalGameManagers file was found at /home/steam/valheim/valheim_server_Data, quitting player!
Failed to initialize player
Caught fatal signal - signo:11 code:1 errno:0 addr:(nil)
Obtained 10 stack frames.
#0 0x007fa768a77730 in funlockfile
#1 0x007fa7695e40e6 in PhysicsModule::GetDefaultPhysicsSceneHandle() const
#2 0x007fa769537c1b in UnityScene::~UnityScene()
#3 0x007fa769535217 in RuntimeSceneManager::~RuntimeSceneManager()
#4 0x007fa769534b15 in StaticDestroyRuntimeSceneManager(void*)
#5 0x007fa76929f42a in RegisterRuntimeInitializeAndCleanup::ExecuteCleanup()
#6 0x007fa7695227b0 in RuntimeCleanup()
#7 0x007fa76968763a in PlayerMain(int, char**)
#8 0x007fa7688c809b in __libc_start_main
#9 0x005608772f8699 in _start
```

By any chance, has anyone ever run into anything similar before?

**Comment by mbround18** [comment]
That's a weird error, are you sure install pulled all the correct files? I'd recommend nuking your server installation and restarting the container.

**Reply by Android3000** [reply]
Thanks for following up. 

You got it, that's exactly what I was indicating happened -- SteamCMD didn't pull all the correct files. It appears to have been missing a file which led to the errors (and as mentioned, the nuke/restart worked fine). It looks like 0x626 hints that it may be in StateFilesMissing based on what I'm seeing here if I'm doing the calculation right: https://docs.linuxgsm.com/steamcmd/errors

I was just curious if others have run into this same thing with any frequency. I haven't really used SteamCMD before to know if it's flaky at times. If so, I wonder if some SteamCMD retry logic for the Odin Install command that the auto_update script invokes might be worth adding in order to make another attempt if there is a non-zero exit code. 

Or maybe this almost never happens, and I just got hit by a very infrequent bug... but figured I'd check :)

**Reply by CosmicHorrorDev** [reply]
That's not an error I've seen before, but I could see the use in having a set number of retries for things like installing. I'm not sure how robust steamcmd is to things like spurious network errors, so it could be good to have!
